// @contentstack/personalize-edge-sdk/dist/personalize.spec.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var cross_fetch_1 = __importDefault(require("cross-fetch"));
var personalize_1 = require("./personalize");
var sdk_1 = require("./sdk");
jest.mock('cross-fetch');
jest.mock('./sdk');
var manifestResponse = {
    experiences: [
        {
            shortUid: '0',
            activeVariantShortUid: '0'
        },
        {
            shortUid: '1',
            activeVariantShortUid: '1'
        },
        {
            shortUid: '2',
            activeVariantShortUid: null
        },
    ]
};
var mockCookie = '';
var xUserAgentHeader = {
    'X-User-Agent': 'personalize-edge-sdk-js/0.0.0-development'
};
function initializeDocument() {
    Object.defineProperty(document, 'cookie', {
        get: jest.fn(function () { return mockCookie; }),
        set: jest.fn(function (newCookie) {
            var cookieName = newCookie.split('=')[0];
            var cookieValue = newCookie.split('=')[1].split(';')[0];
            var existingCookie = mockCookie.split('; ').find(function (cookie) {
                return cookie.split('=')[0] === cookieName;
            });
            if (existingCookie) {
                mockCookie = mockCookie.replace(existingCookie, "".concat(cookieName, "=").concat(cookieValue));
            }
            else {
                mockCookie += (mockCookie ? '; ' : '') + newCookie;
            }
        })
    });
}
describe('initialization', function () {
    var projectUid = 'projectUid';
    var userUid = 'randomly-generated-user-uid';
    var defaultEdgeApiUrl = 'https://personalize-edge.contentstack.com';
    var defaultVariantQueryParam = 'personalize_variants';
    var manifestFetchResult;
    beforeEach(function () {
        personalize_1.Personalize.resetData();
        document = {};
        initializeDocument();
        jest.clearAllMocks();
        manifestFetchResult = {
            ok: true,
            status: 200,
            headers: { get: jest.fn().mockReturnValue(userUid) },
            json: jest.fn().mockResolvedValue(manifestResponse)
        };
        cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
    });
    afterEach(function () {
        mockCookie = '';
    });
    it('initializes the SDK', function () { return __awaiter(void 0, void 0, void 0, function () {
        var result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    expect(personalize_1.Personalize.getInitializationStatus()).toBeUndefined();
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid)];
                case 1:
                    result = _a.sent();
                    expect(personalize_1.Personalize.getInitializationStatus()).toBe('success');
                    expect(cross_fetch_1["default"]).toBeCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid }, xUserAgentHeader)
                    });
                    expect(manifestFetchResult.json).toBeCalled();
                    expect(result).toBeInstanceOf(sdk_1.Sdk);
                    expect(sdk_1.Sdk).toBeCalledWith(projectUid, manifestResponse, defaultEdgeApiUrl, userUid, true, defaultVariantQueryParam);
                    return [2 /*return*/];
            }
        });
    }); });
    it('initializes the SDK in a browser environment', function () { return __awaiter(void 0, void 0, void 0, function () {
        var result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    document.cookie = 'cs-personalize-user-uid=test-user-id';
                    document.URL = 'http://example.com/example?abc=pqr&par%3F%3Dam2=value2%3D';
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid)];
                case 1:
                    result = _a.sent();
                    expect(personalize_1.Personalize.getInitializationStatus()).toBe('success');
                    expect(cross_fetch_1["default"]).toBeCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'test-user-id', 'X-Page-Url': 'http://example.com/example?abc=pqr&par%3F%3Dam2=value2%3D' }, xUserAgentHeader)
                    });
                    expect(manifestFetchResult.json).toBeCalled();
                    expect(result).toBeInstanceOf(sdk_1.Sdk);
                    expect(sdk_1.Sdk).toBeCalledWith(projectUid, manifestResponse, defaultEdgeApiUrl, userUid, true, defaultVariantQueryParam);
                    return [2 /*return*/];
            }
        });
    }); });
    it('initializes the SDK in a non-browser environment', function () { return __awaiter(void 0, void 0, void 0, function () {
        var result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    document = undefined;
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid)];
                case 1:
                    result = _a.sent();
                    expect(personalize_1.Personalize.getInitializationStatus()).toBe('success');
                    expect(cross_fetch_1["default"]).toBeCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid }, xUserAgentHeader)
                    });
                    expect(manifestFetchResult.json).toBeCalled();
                    expect(result).toBeInstanceOf(sdk_1.Sdk);
                    expect(sdk_1.Sdk).toBeCalledWith(projectUid, manifestResponse, defaultEdgeApiUrl, userUid, false, defaultVariantQueryParam);
                    return [2 /*return*/];
            }
        });
    }); });
    it('includes referer in manifest headers if referer exists', function () { return __awaiter(void 0, void 0, void 0, function () {
        var result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    document.referrer = 'https://app.example.com/';
                    document.cookie = 'cs-personalize-user-uid=test-user-id';
                    document.URL = 'http://example.com/example?abc=pqr&par%3F%3Dam2=value2%3D';
                    manifestFetchResult = {
                        ok: true,
                        status: 200,
                        headers: { get: jest.fn().mockReturnValue('test-user-id') },
                        json: jest.fn().mockResolvedValue(manifestResponse)
                    };
                    cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid)];
                case 1:
                    result = _a.sent();
                    expect(personalize_1.Personalize.getInitializationStatus()).toBe('success');
                    expect(cross_fetch_1["default"]).toBeCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'test-user-id', Referer: 'https://app.example.com/', 'X-Page-Url': 'http://example.com/example?abc=pqr&par%3F%3Dam2=value2%3D' }, xUserAgentHeader)
                    });
                    expect(manifestFetchResult.json).toBeCalled();
                    expect(result).toBeInstanceOf(sdk_1.Sdk);
                    expect(sdk_1.Sdk).toBeCalledWith(projectUid, manifestResponse, defaultEdgeApiUrl, 'test-user-id', true, defaultVariantQueryParam);
                    return [2 /*return*/];
            }
        });
    }); });
    it('gets the correct user uid cookie value when multiple cookies present', function () { return __awaiter(void 0, void 0, void 0, function () {
        var result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    document.cookie = 'hello=world; cs-personalize-user-uid=test-user-id1; abc=xyz;';
                    manifestFetchResult = {
                        ok: true,
                        status: 200,
                        headers: { get: jest.fn().mockReturnValue('test-user-id1') },
                        json: jest.fn().mockResolvedValue(manifestResponse)
                    };
                    cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid)];
                case 1:
                    result = _a.sent();
                    expect(cross_fetch_1["default"]).toBeCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'test-user-id1' }, xUserAgentHeader)
                    });
                    expect(result).toBeInstanceOf(sdk_1.Sdk);
                    expect(sdk_1.Sdk).toBeCalledWith(projectUid, manifestResponse, defaultEdgeApiUrl, 'test-user-id1', true, defaultVariantQueryParam);
                    return [2 /*return*/];
            }
        });
    }); });
    it('gets user uid from response header and sets it as a cookie', function () { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, personalize_1.Personalize.init(projectUid)];
                case 1:
                    _a.sent();
                    expect(document.cookie).toBe('cs-personalize-user-uid=randomly-generated-user-uid; Max-Age=31536000; Path=/');
                    return [2 /*return*/];
            }
        });
    }); });
    it('sets the random user id from manifest response in memory', function () { return __awaiter(void 0, void 0, void 0, function () {
        var userId;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    document = undefined;
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid)];
                case 1:
                    _a.sent();
                    expect(personalize_1.Personalize.getInitializationStatus()).toBe('success');
                    userId = personalize_1.Personalize.getUserId();
                    expect(userId).toBe('randomly-generated-user-uid');
                    return [2 /*return*/];
            }
        });
    }); });
    it('with request option, userId is set from request cookie header', function () { return __awaiter(void 0, void 0, void 0, function () {
        var mockRequest, result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    mockRequest = getMockRequest({
                        cookie: 'cs-personalize-user-uid=test-user-id'
                    });
                    manifestFetchResult = {
                        ok: true,
                        status: 200,
                        headers: { get: jest.fn().mockReturnValue('test-user-id') },
                        json: jest.fn().mockResolvedValue(manifestResponse)
                    };
                    cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid, {
                            request: mockRequest
                        })];
                case 1:
                    result = _a.sent();
                    expect(cross_fetch_1["default"]).toBeCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'test-user-id' }, xUserAgentHeader)
                    });
                    expect(result).toBeInstanceOf(sdk_1.Sdk);
                    expect(sdk_1.Sdk).toBeCalledWith(projectUid, manifestResponse, defaultEdgeApiUrl, 'test-user-id', true, defaultVariantQueryParam);
                    return [2 /*return*/];
            }
        });
    }); });
    it('sends manifest request without userId when cookiestring is received as null from request option', function () { return __awaiter(void 0, void 0, void 0, function () {
        var mockRequest, result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    mockRequest = getMockRequest({
                        cookie: null
                    });
                    manifestFetchResult = {
                        ok: true,
                        status: 200,
                        headers: { get: jest.fn().mockReturnValue('test-user-id') },
                        json: jest.fn().mockResolvedValue(manifestResponse)
                    };
                    cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid, {
                            request: mockRequest
                        })];
                case 1:
                    result = _a.sent();
                    expect(cross_fetch_1["default"]).toBeCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid }, xUserAgentHeader)
                    });
                    expect(result).toBeInstanceOf(sdk_1.Sdk);
                    expect(sdk_1.Sdk).toBeCalledWith(projectUid, manifestResponse, defaultEdgeApiUrl, 'test-user-id', true, defaultVariantQueryParam);
                    return [2 /*return*/];
            }
        });
    }); });
    it('userId option takes precedence over request cookie userId', function () { return __awaiter(void 0, void 0, void 0, function () {
        var mockRequest, mockUserId, result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    document.cookie = 'cs-personalize-user-uid=test-user-id';
                    mockRequest = getMockRequest({
                        cookie: 'cs-personalize-user-uid=test-user-id'
                    });
                    mockUserId = 'test-user-id1';
                    manifestFetchResult = {
                        ok: true,
                        status: 200,
                        headers: { get: jest.fn().mockReturnValue('test-user-id') },
                        json: jest.fn().mockResolvedValue(manifestResponse)
                    };
                    cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid, {
                            request: mockRequest,
                            userId: mockUserId
                        })];
                case 1:
                    result = _a.sent();
                    expect(cross_fetch_1["default"]).toBeCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': mockUserId }, xUserAgentHeader)
                    });
                    expect(result).toBeInstanceOf(sdk_1.Sdk);
                    expect(sdk_1.Sdk).toBeCalledWith(projectUid, manifestResponse, defaultEdgeApiUrl, 'test-user-id', true, defaultVariantQueryParam);
                    return [2 /*return*/];
            }
        });
    }); });
    it('fetches manifest if manifest cookie value is not present', function () { return __awaiter(void 0, void 0, void 0, function () {
        var mockRequest, result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    document.cookie = 'cs-personalize-user-uid=test-user-id';
                    document.URL = 'http://example.com/example?abc=pqr&par%3F%3Dam2=value2%3D';
                    mockRequest = getMockRequest({
                        cookie: 'cs-personalize-user-uid=test-user-id'
                    });
                    manifestFetchResult = {
                        ok: true,
                        status: 200,
                        headers: { get: jest.fn().mockReturnValue('test-user-id') },
                        json: jest.fn().mockResolvedValue(manifestResponse)
                    };
                    cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid, {
                            request: mockRequest
                        })];
                case 1:
                    result = _a.sent();
                    expect(cross_fetch_1["default"]).toHaveBeenCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'test-user-id', 'X-Page-Url': 'http://example.com/example?abc=pqr&par%3F%3Dam2=value2%3D' }, xUserAgentHeader)
                    });
                    expect(manifestFetchResult.json).toBeCalled();
                    expect(result).toBeInstanceOf(sdk_1.Sdk);
                    expect(sdk_1.Sdk).toBeCalledWith(projectUid, manifestResponse, defaultEdgeApiUrl, 'test-user-id', true, defaultVariantQueryParam);
                    return [2 /*return*/];
            }
        });
    }); });
    it('fetches manifest with lytics audiences live attribute if present in request headers', function () { return __awaiter(void 0, void 0, void 0, function () {
        var request, personalizeSdk;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    manifestFetchResult = {
                        ok: true,
                        status: 200,
                        headers: { get: jest.fn().mockReturnValue('test-user-id') },
                        json: jest.fn().mockResolvedValue(manifestResponse)
                    };
                    cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                    request = getMockRequest({
                        cookie: 'cs-lytics-audiences=|ly_aud_1|ly_aud_2|; cs-personalize-user-uid=test-user-id'
                    });
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid, { request: request })];
                case 1:
                    personalizeSdk = _a.sent();
                    expect(cross_fetch_1["default"]).toHaveBeenCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'test-user-id', 'X-Live-Attributes': '{"__lytics_audiences":"|ly_aud_1|ly_aud_2|"}' }, xUserAgentHeader)
                    });
                    expect(manifestFetchResult.json).toBeCalled();
                    expect(personalizeSdk).toBeInstanceOf(sdk_1.Sdk);
                    return [2 /*return*/];
            }
        });
    }); });
    it('fetches manifest with custom live attributes passed in options', function () { return __awaiter(void 0, void 0, void 0, function () {
        var request, personalizeSdk;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    manifestFetchResult = {
                        ok: true,
                        status: 200,
                        headers: { get: jest.fn().mockReturnValue('test-user-id') },
                        json: jest.fn().mockResolvedValue(manifestResponse)
                    };
                    cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                    request = getMockRequest({
                        cookie: 'cs-personalize-user-uid=test-user-id'
                    });
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid, {
                            liveAttributes: {
                                custom_live_attribute: 'custom_live_attribute_value'
                            },
                            request: request
                        })];
                case 1:
                    personalizeSdk = _a.sent();
                    expect(cross_fetch_1["default"]).toHaveBeenCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'test-user-id', 'X-Live-Attributes': '{"custom_live_attribute":"custom_live_attribute_value"}' }, xUserAgentHeader)
                    });
                    expect(manifestFetchResult.json).toBeCalled();
                    expect(personalizeSdk).toBeInstanceOf(sdk_1.Sdk);
                    return [2 /*return*/];
            }
        });
    }); });
    it('fetches manifest with custom attributes and lytics audiences passed as live attributes', function () { return __awaiter(void 0, void 0, void 0, function () {
        var request, personalizeSdk;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    manifestFetchResult = {
                        ok: true,
                        status: 200,
                        headers: { get: jest.fn().mockReturnValue('test-user-id') },
                        json: jest.fn().mockResolvedValue(manifestResponse)
                    };
                    cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                    request = getMockRequest({
                        cookie: 'cs-lytics-audiences=|ly_aud_1|ly_aud_2|; cs-personalize-user-uid=test-user-id'
                    });
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid, {
                            liveAttributes: {
                                custom_live_attribute: 'custom_live_attribute_value'
                            },
                            request: request
                        })];
                case 1:
                    personalizeSdk = _a.sent();
                    expect(cross_fetch_1["default"]).toHaveBeenCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'test-user-id', 'X-Live-Attributes': '{"__lytics_audiences":"|ly_aud_1|ly_aud_2|","custom_live_attribute":"custom_live_attribute_value"}' }, xUserAgentHeader)
                    });
                    expect(manifestFetchResult.json).toBeCalled();
                    expect(personalizeSdk).toBeInstanceOf(sdk_1.Sdk);
                    return [2 /*return*/];
            }
        });
    }); });
    it('fetches manifest with lytics audiences live attribute if present in cookies when in browser', function () { return __awaiter(void 0, void 0, void 0, function () {
        var personalizeSdk;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    document.cookie = 'cs-personalize-user-uid=test-user-id; cs-lytics-audiences=|ly_aud_1|ly_aud_2|';
                    document.URL = 'http://example.com/example?abc=pqr&par%3F%3Dam2=value2%3D';
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid)];
                case 1:
                    personalizeSdk = _a.sent();
                    expect(cross_fetch_1["default"]).toHaveBeenCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'test-user-id', 'X-Live-Attributes': '{"__lytics_audiences":"|ly_aud_1|ly_aud_2|"}', 'X-Page-Url': 'http://example.com/example?abc=pqr&par%3F%3Dam2=value2%3D' }, xUserAgentHeader)
                    });
                    expect(manifestFetchResult.json).toBeCalled();
                    expect(personalizeSdk).toBeInstanceOf(sdk_1.Sdk);
                    return [2 /*return*/];
            }
        });
    }); });
    it('does not fetch manifest if manifest cookie value is present with older cookie format', function () { return __awaiter(void 0, void 0, void 0, function () {
        var mockRequest;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    document.cookie =
                        'cs-personalize-user-uid=test-user-id; ' +
                            'cs-personalize-manifest={ "activeVariants": { "0": "0", "1": "1", "2": null } }';
                    document.URL = 'http://example.com/example?abc=pqr&par%3F%3Dam2=value2%3D';
                    mockRequest = getMockRequest({});
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid, {
                            request: mockRequest
                        })];
                case 1:
                    _a.sent();
                    expect(cross_fetch_1["default"]).not.toBeCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'test-user-id', 'X-Page-Url': 'http://example.com/example?abc=pqr&par%3F%3Dam2=value2%3D' }, xUserAgentHeader)
                    });
                    return [2 /*return*/];
            }
        });
    }); });
    it('does not fetch manifest if manifest cookie value is present with newer cookie format', function () { return __awaiter(void 0, void 0, void 0, function () {
        var mockRequest;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    document.cookie =
                        'cs-personalize-user-uid=test-user-id; ' + "cs-personalize-manifest=".concat(JSON.stringify(manifestResponse));
                    document.URL = 'http://example.com/example?abc=pqr&par%3F%3Dam2=value2%3D';
                    mockRequest = getMockRequest({});
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid, {
                            request: mockRequest
                        })];
                case 1:
                    _a.sent();
                    expect(cross_fetch_1["default"]).not.toBeCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'test-user-id', 'X-Page-Url': 'http://example.com/example?abc=pqr&par%3F%3Dam2=value2%3D' }, xUserAgentHeader)
                    });
                    return [2 /*return*/];
            }
        });
    }); });
    it("includes x-forwarded-for, referer and x-page-url in manifest headers,\n        if exists in the request option in non-browser environment", function () { return __awaiter(void 0, void 0, void 0, function () {
        var mockRequest, result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    document = undefined;
                    mockRequest = getMockRequest({
                        cookie: 'cs-personalize-user-uid=test-user-id',
                        referer: 'https://app.example.com/',
                        'x-forwarded-for': '127.0.0.1'
                    }, 'http://example.com/example?abc=pqr&par%3F%3Dam2=value2%3D');
                    manifestFetchResult = {
                        ok: true,
                        status: 200,
                        headers: { get: jest.fn().mockReturnValue('test-user-id') },
                        json: jest.fn().mockResolvedValue(manifestResponse)
                    };
                    cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid, {
                            request: mockRequest
                        })];
                case 1:
                    result = _a.sent();
                    expect(cross_fetch_1["default"]).toHaveBeenCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'test-user-id', 'X-Page-Url': 'http://example.com/example?abc=pqr&par%3F%3Dam2=value2%3D', 'X-Forwarded-For': '127.0.0.1', Referer: 'https://app.example.com/' }, xUserAgentHeader)
                    });
                    expect(result).toBeInstanceOf(sdk_1.Sdk);
                    expect(sdk_1.Sdk).toBeCalledWith(projectUid, manifestResponse, defaultEdgeApiUrl, 'test-user-id', false, defaultVariantQueryParam);
                    return [2 /*return*/];
            }
        });
    }); });
    it('includes x-forwarded-for in manifest headers if exists in the request option', function () { return __awaiter(void 0, void 0, void 0, function () {
        var mockRequest, result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    mockRequest = getMockRequest({ 'x-forwarded-for': '127.0.0.1' });
                    manifestFetchResult = {
                        ok: true,
                        status: 200,
                        headers: { get: jest.fn().mockReturnValue('test-user-id') },
                        json: jest.fn().mockResolvedValue(manifestResponse)
                    };
                    cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid, {
                            request: mockRequest
                        })];
                case 1:
                    result = _a.sent();
                    expect(cross_fetch_1["default"]).toHaveBeenCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'test-user-id', 'X-Forwarded-For': '127.0.0.1' }, xUserAgentHeader)
                    });
                    expect(result).toBeInstanceOf(sdk_1.Sdk);
                    expect(sdk_1.Sdk).toBeCalledWith(projectUid, manifestResponse, defaultEdgeApiUrl, 'test-user-id', true, defaultVariantQueryParam);
                    return [2 /*return*/];
            }
        });
    }); });
    it('includes referrer in manifest headers if exists in the request option', function () { return __awaiter(void 0, void 0, void 0, function () {
        var mockRequest, result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    document.referrer = 'https://app.example123.com/';
                    document.cookie = 'cs-personalize-user-uid=test-user-id';
                    mockRequest = getMockRequest({
                        referer: 'https://app.example.com/'
                    });
                    manifestFetchResult = {
                        ok: true,
                        status: 200,
                        headers: { get: jest.fn().mockReturnValue('test-user-id') },
                        json: jest.fn().mockResolvedValue(manifestResponse)
                    };
                    cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid, {
                            request: mockRequest
                        })];
                case 1:
                    result = _a.sent();
                    expect(cross_fetch_1["default"]).toHaveBeenCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'test-user-id', Referer: 'https://app.example.com/' }, xUserAgentHeader)
                    });
                    expect(result).toBeInstanceOf(sdk_1.Sdk);
                    expect(sdk_1.Sdk).toBeCalledWith(projectUid, manifestResponse, defaultEdgeApiUrl, 'test-user-id', true, defaultVariantQueryParam);
                    return [2 /*return*/];
            }
        });
    }); });
    it('includes x-page-url in manifest headers if exists in the request', function () { return __awaiter(void 0, void 0, void 0, function () {
        var mockRequest, result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    document.cookie = 'cs-personalize-user-uid=test-user-id';
                    document.URL = 'http://example.com/example123?abc=pqr&par%3F%3Dam2=value2%3D';
                    mockRequest = getMockRequest({}, 'http://example.com/example?abc=pqr&par%3F%3Dam2=value2%3D');
                    manifestFetchResult = {
                        ok: true,
                        status: 200,
                        headers: { get: jest.fn().mockReturnValue('test-user-id') },
                        json: jest.fn().mockResolvedValue(manifestResponse)
                    };
                    cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid, {
                            request: mockRequest
                        })];
                case 1:
                    result = _a.sent();
                    expect(cross_fetch_1["default"]).toHaveBeenCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'test-user-id', 'X-Page-Url': 'http://example.com/example?abc=pqr&par%3F%3Dam2=value2%3D' }, xUserAgentHeader)
                    });
                    expect(result).toBeInstanceOf(sdk_1.Sdk);
                    expect(sdk_1.Sdk).toBeCalledWith(projectUid, manifestResponse, defaultEdgeApiUrl, 'test-user-id', true, defaultVariantQueryParam);
                    return [2 /*return*/];
            }
        });
    }); });
    it('includes user-agent in manifest headers if exists in the request', function () { return __awaiter(void 0, void 0, void 0, function () {
        var userAgent, mockRequest, result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    document.cookie = 'cs-personalize-user-uid=test-user-id';
                    userAgent = "Mozilla/5.0 (Macintosh;\n       Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36";
                    mockRequest = getMockRequest({
                        'user-agent': userAgent
                    });
                    manifestFetchResult = {
                        ok: true,
                        status: 200,
                        headers: { get: jest.fn().mockReturnValue('test-user-id') },
                        json: jest.fn().mockResolvedValue(manifestResponse)
                    };
                    cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                    return [4 /*yield*/, personalize_1.Personalize.init(projectUid, {
                            request: mockRequest
                        })];
                case 1:
                    result = _a.sent();
                    expect(cross_fetch_1["default"]).toHaveBeenCalledWith('https://personalize-edge.contentstack.com/manifest', {
                        headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'test-user-id', 'User-Agent': userAgent }, xUserAgentHeader)
                    });
                    expect(result).toBeInstanceOf(sdk_1.Sdk);
                    expect(sdk_1.Sdk).toBeCalledWith(projectUid, manifestResponse, defaultEdgeApiUrl, 'test-user-id', true, defaultVariantQueryParam);
                    return [2 /*return*/];
            }
        });
    }); });
    it('throws a helpful error if manifest api returns project not found response', function () { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    manifestFetchResult = {
                        status: 404,
                        ok: false,
                        json: jest.fn().mockResolvedValue({ error: 'personalize.MANIFEST.PROJECT_NOT_FOUND', message: 'Not found' })
                    };
                    cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                    return [4 /*yield*/, expect(personalize_1.Personalize.init(projectUid)).rejects.toThrow("Project not found: ".concat(projectUid))];
                case 1:
                    _a.sent();
                    expect(personalize_1.Personalize.getInitializationStatus()).toBe('error');
                    return [2 /*return*/];
            }
        });
    }); });
    it.each([
        [400, 'Bad request'],
        [404, 'Unknown not found error'],
        [500, 'Internal server error'],
        [502, 'Bad gateway'],
        [503, 'Service unavailable'],
    ])('throws an error if manifest api returns an unknown error with status %i', function (status, message) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    manifestFetchResult = {
                        status: status,
                        ok: false,
                        json: jest.fn().mockResolvedValue({ error: 'unknown', message: message })
                    };
                    cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                    return [4 /*yield*/, expect(personalize_1.Personalize.init(projectUid)).rejects.toThrow("Could not fetch manifest: ".concat(status))];
                case 1:
                    _a.sent();
                    expect(personalize_1.Personalize.getInitializationStatus()).toBe('error');
                    return [2 /*return*/];
            }
        });
    }); });
});
describe('SDK', function () {
    var projectUid = 'projectUid';
    var userUid = 'randomly-generated-user-uid';
    var manifestFetchResult;
    describe('global methods', function () {
        var sdk;
        beforeEach(function () { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        personalize_1.Personalize.resetData();
                        jest.clearAllMocks();
                        document = {};
                        initializeDocument();
                        manifestFetchResult = {
                            ok: true,
                            status: 200,
                            headers: { get: jest.fn().mockReturnValue(userUid) },
                            json: jest.fn().mockResolvedValue(manifestResponse)
                        };
                        cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                        sdk_1.Sdk.mockReturnValue({
                            getActiveVariant: jest.fn().mockReturnValue('0'),
                            getVariants: jest.fn().mockReturnValue({ '0': '1', '1': '2' }),
                            getExperiences: jest.fn().mockReturnValue(manifestResponse.experiences),
                            triggerImpression: jest.fn(),
                            triggerImpressions: jest.fn(),
                            triggerEvent: jest.fn(),
                            setUserId: jest.fn(),
                            getUserId: jest.fn(),
                            set: jest.fn(),
                            addStateToResponse: jest.fn(),
                            getVariantAliases: jest
                                .fn()
                                .mockReturnValue(['cs_personalize_0_0', 'cs_personalize_1_1']),
                            getVariantParam: jest.fn().mockReturnValue('a_0')
                        });
                        return [4 /*yield*/, personalize_1.Personalize.init(projectUid)];
                    case 1:
                        sdk = _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
        describe('set user id', function () {
            it('sets user id as a cookie', function () { return __awaiter(void 0, void 0, void 0, function () {
                var mockCustomUserUid;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            mockCustomUserUid = 'mock-custom-user-uid';
                            return [4 /*yield*/, personalize_1.Personalize.setUserId(mockCustomUserUid)];
                        case 1:
                            _a.sent();
                            expect(sdk.setUserId).toHaveBeenCalledWith(mockCustomUserUid, undefined);
                            return [2 /*return*/];
                    }
                });
            }); });
            it('sets user id when sdk is not initialized', function () { return __awaiter(void 0, void 0, void 0, function () {
                var mockCustomUserUid;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            personalize_1.Personalize.resetData();
                            mockCustomUserUid = 'mock-custom-user-uid';
                            return [4 /*yield*/, personalize_1.Personalize.setUserId(mockCustomUserUid)];
                        case 1:
                            _a.sent();
                            expect(personalize_1.Personalize.getUserId()).toBe(mockCustomUserUid);
                            return [2 /*return*/];
                    }
                });
            }); });
        });
        describe('get user id', function () {
            it('returns the user id from memory if present', function () { return __awaiter(void 0, void 0, void 0, function () {
                var mockCustomUserUid, userId;
                return __generator(this, function (_a) {
                    mockCustomUserUid = 'mock-custom-user-uid';
                    personalize_1.Personalize.resetData();
                    personalize_1.Personalize.setUserId(mockCustomUserUid);
                    userId = personalize_1.Personalize.getUserId();
                    expect(userId).toBe(mockCustomUserUid);
                    return [2 /*return*/];
                });
            }); });
            it('returns undefined if no user id present in memory', function () { return __awaiter(void 0, void 0, void 0, function () {
                var userId;
                return __generator(this, function (_a) {
                    personalize_1.Personalize.resetData();
                    userId = personalize_1.Personalize.getUserId();
                    expect(userId).toBe(undefined);
                    return [2 /*return*/];
                });
            }); });
        });
        describe('set user attributes', function () {
            it('sets the user attributes', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, personalize_1.Personalize.set({
                                name: 'Eclipse'
                            })];
                        case 1:
                            _a.sent();
                            expect(sdk.set).toHaveBeenCalledWith({
                                name: 'Eclipse'
                            });
                            return [2 /*return*/];
                    }
                });
            }); });
            it('waits for initialization to succeed before setting attributes', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            personalize_1.Personalize.init(projectUid);
                            expect(personalize_1.Personalize.getInitializationStatus()).toBe('initializing');
                            return [4 /*yield*/, personalize_1.Personalize.set({
                                    name: 'Eclipse'
                                })];
                        case 1:
                            _a.sent();
                            expect(sdk.set).toHaveBeenCalledWith({
                                name: 'Eclipse'
                            });
                            return [2 /*return*/];
                    }
                });
            }); });
            it('throws error if called before initialization', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    personalize_1.Personalize.resetData();
                    expect(function () { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, personalize_1.Personalize.set({
                                        name: 'Eclipse'
                                    })];
                                case 1:
                                    _a.sent();
                                    return [2 /*return*/];
                            }
                        });
                    }); }).rejects.toThrow('Cannot set user attributes before initialization');
                    return [2 /*return*/];
                });
            }); });
        });
        describe('get active variant from experience', function () {
            it('returns active variant short uid', function () { return __awaiter(void 0, void 0, void 0, function () {
                var expectedShortUid, result;
                return __generator(this, function (_a) {
                    expectedShortUid = '0';
                    result = personalize_1.Personalize.getActiveVariant('0');
                    expect(sdk.getActiveVariant).toHaveBeenCalledWith('0');
                    expect(result).toBe(expectedShortUid);
                    return [2 /*return*/];
                });
            }); });
            it('throws error if called before initialization', function () {
                personalize_1.Personalize.resetData();
                expect(function () { return __awaiter(void 0, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        personalize_1.Personalize.getActiveVariant('0');
                        return [2 /*return*/];
                    });
                }); }).rejects.toThrow('Cannot get active variant before initialization');
            });
        });
        describe('trigger impressions', function () {
            it('triggers the impression', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, personalize_1.Personalize.triggerImpression('1')];
                        case 1:
                            _a.sent();
                            expect(sdk.triggerImpression).toHaveBeenCalledWith('1');
                            return [2 /*return*/];
                    }
                });
            }); });
            it('triggers multiple impressions', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, personalize_1.Personalize.triggerImpressions({ experienceShortUids: ['0', '1'] })];
                        case 1:
                            _a.sent();
                            expect(sdk.triggerImpressions).toHaveBeenCalledWith({ experienceShortUids: ['0', '1'] });
                            return [2 /*return*/];
                    }
                });
            }); });
            it('triggerImpressions throws error if called before initialization', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    personalize_1.Personalize.resetData();
                    expect(function () { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, personalize_1.Personalize.triggerImpressions({
                                        experienceShortUids: ['0']
                                    })];
                                case 1:
                                    _a.sent();
                                    return [2 /*return*/];
                            }
                        });
                    }); }).rejects.toThrow('Cannot trigger impressions before initialization');
                    return [2 /*return*/];
                });
            }); });
            it('waits for initialization to succeed before triggering impressions', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            personalize_1.Personalize.init(projectUid);
                            expect(personalize_1.Personalize.getInitializationStatus()).toBe('initializing');
                            return [4 /*yield*/, personalize_1.Personalize.triggerImpressions({ experienceShortUids: ['1'] })];
                        case 1:
                            _a.sent();
                            expect(sdk.triggerImpressions).toHaveBeenCalledWith({ experienceShortUids: ['1'] });
                            return [2 /*return*/];
                    }
                });
            }); });
            it('triggerImpression throws error if called before initialization', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    personalize_1.Personalize.resetData();
                    expect(function () { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, personalize_1.Personalize.triggerImpression('0')];
                                case 1:
                                    _a.sent();
                                    return [2 /*return*/];
                            }
                        });
                    }); }).rejects.toThrow('Cannot trigger impression before initialization');
                    return [2 /*return*/];
                });
            }); });
            it('waits for initialization to succeed before triggering impression', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            personalize_1.Personalize.init(projectUid);
                            expect(personalize_1.Personalize.getInitializationStatus()).toBe('initializing');
                            return [4 /*yield*/, personalize_1.Personalize.triggerImpression('1')];
                        case 1:
                            _a.sent();
                            expect(sdk.triggerImpression).toHaveBeenCalledWith('1');
                            return [2 /*return*/];
                    }
                });
            }); });
            it('triggers impressions with aliases', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, personalize_1.Personalize.triggerImpressions({ aliases: personalize_1.Personalize.getVariantAliases() })];
                        case 1:
                            _a.sent();
                            expect(sdk.triggerImpressions).toHaveBeenCalledWith({ aliases: personalize_1.Personalize.getVariantAliases() });
                            return [2 /*return*/];
                    }
                });
            }); });
            it('throws an error if neither aliases nor shortUids are provided', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    expect(function () { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, personalize_1.Personalize.triggerImpressions({})];
                                case 1:
                                    _a.sent();
                                    return [2 /*return*/];
                            }
                        });
                    }); }).rejects.toThrow('Cannot trigger impressions without options');
                    return [2 /*return*/];
                });
            }); });
        });
        describe('A/B Test', function () {
            describe('trigger event', function () {
                it('triggers an event', function () { return __awaiter(void 0, void 0, void 0, function () {
                    var eventKey;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                document.cookie = 'cs-personalize-user-uid=test-user-id';
                                eventKey = 'event_1';
                                return [4 /*yield*/, personalize_1.Personalize.triggerEvent(eventKey)];
                            case 1:
                                _a.sent();
                                expect(sdk.triggerEvent).toHaveBeenCalledWith(eventKey);
                                return [2 /*return*/];
                        }
                    });
                }); });
                it('throws error if called before initialization', function () { return __awaiter(void 0, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        personalize_1.Personalize.resetData();
                        expect(function () { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, personalize_1.Personalize.triggerEvent('event_key')];
                                    case 1:
                                        _a.sent();
                                        return [2 /*return*/];
                                }
                            });
                        }); }).rejects.toThrow('Cannot trigger event before initialization');
                        return [2 /*return*/];
                    });
                }); });
                it('waits for initialization to succeed before sending the event', function () { return __awaiter(void 0, void 0, void 0, function () {
                    var eventKey;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                document.cookie = 'cs-personalize-user-uid=test-user-id';
                                eventKey = 'event_1';
                                personalize_1.Personalize.init(projectUid);
                                expect(personalize_1.Personalize.getInitializationStatus()).toBe('initializing');
                                return [4 /*yield*/, personalize_1.Personalize.triggerEvent(eventKey)];
                            case 1:
                                _a.sent();
                                expect(sdk.triggerEvent).toHaveBeenCalledWith(eventKey);
                                return [2 /*return*/];
                        }
                    });
                }); });
            });
        });
        describe('add state to response', function () {
            it('sets the userId and manifest in the set-cookie header', function () { return __awaiter(void 0, void 0, void 0, function () {
                var mockResponse;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            mockResponse = {
                                headers: {
                                    set: jest.fn(),
                                    append: jest.fn()
                                }
                            };
                            return [4 /*yield*/, personalize_1.Personalize.init(projectUid)];
                        case 1:
                            _a.sent();
                            return [4 /*yield*/, personalize_1.Personalize.addStateToResponse(mockResponse)];
                        case 2:
                            _a.sent();
                            expect(sdk.addStateToResponse).toHaveBeenCalledWith(mockResponse);
                            return [2 /*return*/];
                    }
                });
            }); });
            it('throws error if called before initialization', function () { return __awaiter(void 0, void 0, void 0, function () {
                var mockResponse;
                return __generator(this, function (_a) {
                    mockResponse = {
                        headers: {
                            set: jest.fn()
                        }
                    };
                    personalize_1.Personalize.resetData();
                    expect(function () { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, personalize_1.Personalize.addStateToResponse(mockResponse)];
                                case 1:
                                    _a.sent();
                                    return [2 /*return*/];
                            }
                        });
                    }); }).rejects.toThrow('Cannot add state to response before initialization');
                    return [2 /*return*/];
                });
            }); });
        });
        describe('get variants', function () {
            it('returns the active variants from manifest', function () { return __awaiter(void 0, void 0, void 0, function () {
                var result;
                return __generator(this, function (_a) {
                    result = personalize_1.Personalize.getVariants();
                    expect(sdk.getVariants).toHaveBeenCalled();
                    expect(result).toStrictEqual({
                        '0': '1',
                        '1': '2'
                    });
                    return [2 /*return*/];
                });
            }); });
            it('throws error if called before initialization', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    personalize_1.Personalize.resetData();
                    expect(function () { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            personalize_1.Personalize.getVariants();
                            return [2 /*return*/];
                        });
                    }); }).rejects.toThrow('Cannot get variants before initialization');
                    return [2 /*return*/];
                });
            }); });
        });
        describe('get experiences', function () {
            it('returns the active experiences', function () { return __awaiter(void 0, void 0, void 0, function () {
                var result;
                return __generator(this, function (_a) {
                    result = personalize_1.Personalize.getExperiences();
                    expect(result).toStrictEqual(manifestResponse.experiences);
                    return [2 /*return*/];
                });
            }); });
            it('throws error if called before initialization', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    personalize_1.Personalize.resetData();
                    expect(function () { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            personalize_1.Personalize.getExperiences();
                            return [2 /*return*/];
                        });
                    }); }).rejects.toThrow('Cannot get experiences before initialization');
                    return [2 /*return*/];
                });
            }); });
        });
        describe('alias management', function () {
            it('returns variant aliases', function () { return __awaiter(void 0, void 0, void 0, function () {
                var aliases;
                return __generator(this, function (_a) {
                    aliases = personalize_1.Personalize.getVariantAliases();
                    expect(sdk.getVariantAliases).toHaveBeenCalled();
                    expect(aliases).toEqual(['cs_personalize_0_0', 'cs_personalize_1_1']);
                    return [2 /*return*/];
                });
            }); });
            it('throws error if called before initialization', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    personalize_1.Personalize.resetData();
                    expect(function () { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            personalize_1.Personalize.getVariantAliases();
                            return [2 /*return*/];
                        });
                    }); }).rejects.toThrow('Cannot get variant aliases before initialization');
                    return [2 /*return*/];
                });
            }); });
            it('encodes aliases as a compact query parameter', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, personalize_1.Personalize.init(projectUid)];
                        case 1:
                            _a.sent();
                            personalize_1.Personalize.getVariantParam();
                            expect(sdk.getVariantParam).toHaveBeenCalled();
                            return [2 /*return*/];
                    }
                });
            }); });
            it('throws error if called before initialization', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    personalize_1.Personalize.resetData();
                    expect(function () { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            personalize_1.Personalize.getVariantParam();
                            return [2 /*return*/];
                        });
                    }); }).rejects.toThrow('Cannot get variant alias param before initialization');
                    return [2 /*return*/];
                });
            }); });
            it('converts variant param to aliases', function () {
                var variantParam = '0_0,1_1';
                expect(personalize_1.Personalize.variantParamToVariantAliases(variantParam)).toEqual([
                    'cs_personalize_0_0',
                    'cs_personalize_1_1',
                ]);
            });
            it('allows customizing the query parameter', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            expect(personalize_1.Personalize.VARIANT_QUERY_PARAM).toEqual('personalize_variants');
                            return [4 /*yield*/, personalize_1.Personalize.init(projectUid, { customVariantQueryParam: 'variants' })];
                        case 1:
                            _a.sent();
                            expect(personalize_1.Personalize.VARIANT_QUERY_PARAM).toEqual('variants');
                            return [2 /*return*/];
                    }
                });
            }); });
        });
        describe('reset', function () {
            it('resets the user id & manifest in non browser environment', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            document = undefined;
                            return [4 /*yield*/, personalize_1.Personalize.init(projectUid)];
                        case 1:
                            _a.sent();
                            expect(personalize_1.Personalize.getUserId()).toBeDefined();
                            expect(personalize_1.Personalize.getVariants()).toBeDefined();
                            personalize_1.Personalize.reset();
                            expect(personalize_1.Personalize.getInitializationStatus()).toBeUndefined();
                            expect(personalize_1.Personalize.getUserId()).toBeUndefined();
                            expect(function () { return __awaiter(void 0, void 0, void 0, function () {
                                return __generator(this, function (_a) {
                                    personalize_1.Personalize.getVariants();
                                    return [2 /*return*/];
                                });
                            }); }).rejects.toThrow('Cannot get variants before initialization');
                            return [2 /*return*/];
                    }
                });
            }); });
            it('clears the user id & manifest cookie in the browser environment', function () { return __awaiter(void 0, void 0, void 0, function () {
                var mockUserId;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            mockUserId = 'test-user-id';
                            document.cookie =
                                'cs-personalize-user-uid=test-user-id; ' + 'cs-personalize-manifest={"activeVariants":{"0":"0","1":"1"}};';
                            manifestFetchResult = {
                                ok: true,
                                status: 200,
                                headers: { get: jest.fn().mockReturnValue(mockUserId) },
                                json: jest.fn().mockResolvedValue(manifestResponse)
                            };
                            cross_fetch_1["default"].mockResolvedValue(manifestFetchResult);
                            return [4 /*yield*/, personalize_1.Personalize.init(projectUid)];
                        case 1:
                            _a.sent();
                            expect(personalize_1.Personalize.getUserId()).toBe(mockUserId);
                            expect(personalize_1.Personalize.getVariants()).toStrictEqual({
                                '0': '1',
                                '1': '2'
                            });
                            personalize_1.Personalize.reset();
                            expect(personalize_1.Personalize.getUserId()).toBeUndefined();
                            expect(function () { return __awaiter(void 0, void 0, void 0, function () {
                                return __generator(this, function (_a) {
                                    personalize_1.Personalize.getVariants();
                                    return [2 /*return*/];
                                });
                            }); }).rejects.toThrow('Cannot get variants before initialization');
                            expect(document.cookie).toBe('cs-personalize-user-uid=; Max-Age=31536000; Path=/; cs-personalize-manifest=;' +
                                ' expires=Thu, 01 Jan 1970 00:00:00 GMT; path=/');
                            return [2 /*return*/];
                    }
                });
            }); });
        });
        describe('getInitializationStatus', function () {
            it('returns undefined when SDK is not initialized', function () {
                personalize_1.Personalize.resetData();
                expect(personalize_1.Personalize.getInitializationStatus()).toBeUndefined();
            });
            it('returns initializing when called while SDK is being initialized', function () {
                personalize_1.Personalize.init(projectUid);
                expect(personalize_1.Personalize.getInitializationStatus()).toBe('initializing');
            });
            it('returns success when called while SDK is initialized', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, personalize_1.Personalize.init(projectUid)];
                        case 1:
                            _a.sent();
                            expect(personalize_1.Personalize.getInitializationStatus()).toBe('success');
                            return [2 /*return*/];
                    }
                });
            }); });
        });
    });
});
function getMockRequest(headers, url) {
    return {
        headers: {
            has: jest.fn(function (header) { return headers.hasOwnProperty(header.toLowerCase()) && headers[header.toLowerCase()] !== null; }),
            get: jest.fn(function (header) {
                if (headers.hasOwnProperty(header.toLowerCase())) {
                    return headers[header.toLowerCase()];
                }
                else {
                    return 'cs-personalize-user-uid=test-user-id';
                }
            })
        },
        url: url
    };
}
