// @contentstack/personalize-edge-sdk/dist/sdk.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.Sdk = void 0;
var cross_fetch_1 = __importDefault(require("cross-fetch"));
var build_info_1 = __importDefault(require("./build-info"));
var event_ingestion_type_1 = require("./event-ingestion.type");
var PROJECT_UID_HEADER_KEY = 'X-Project-Uid';
var USER_UID_HEADER_KEY = 'X-CS-Personalize-User-Uid';
var USER_UID_COOKIE_KEY = 'cs-personalize-user-uid';
var USER_UID_COOKIE_MAX_AGE = 365;
var MANIFEST_COOKIE_KEY = 'cs-personalize-manifest';
var SET_COOKIE_HEADER_KEY = 'Set-Cookie';
var CS_PERSONALIZE_PREFIX = 'cs_personalize';
var X_USER_AGENT = 'X-User-Agent';
var X_USER_AGENT_VALUE = "personalize-edge-sdk-js/".concat(build_info_1["default"].version);
var Sdk = /** @class */ (function () {
    function Sdk(projectUid, manifestData, edgeApiURL, userId, isBrowserEnvironment, variantQueryParam) {
        this._isBrowserEnvironment = false;
        this._projectUid = projectUid;
        this._manifestData = manifestData;
        this._edgeApiURL = edgeApiURL;
        this._userId = userId;
        this._isBrowserEnvironment = isBrowserEnvironment;
        this.VARIANT_QUERY_PARAM = variantQueryParam;
    }
    /**
     * Gets the active experiences
     *
     * Returns a list of experience and corresponding active variant short UIDs.
     *
     * An inactive experience will have `null` in the corresponding `activeVariantShortUid`.
     *
     * The list is sorted in the order of experience priority. The experiences higher in the
     * list have higher priority.
     *
     * Call this method after initializing the SDK.
     *
     * @returns The active experiences
     * @example
     * ```ts
     * sdk.getExperiences(); // [{shortUid: 'a', activeVariantShortUid: '0'}]
     * ```
     */
    Sdk.prototype.getExperiences = function () {
        return this._manifestData.experiences;
    };
    /**
     * Triggers an impression
     *
     * The `triggerImpression` method triggers an impression for a given experience. The impression
     * is sent for the active variant in the manifest. Call this method whenever an experience is
     * shown to the user.
     *
     * Call this method after initializing the SDK.
     *
     * @param experienceShortUid - The experience short UID to trigger the impression for.
     * @example
     * ```ts
     * const experienceShortUid = 'a';
     * await sdk.triggerImpression(experienceShortUid);
     * ```
     */
    Sdk.prototype.triggerImpression = function (experienceShortUid) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.triggerImpressions({
                        experienceShortUids: [experienceShortUid]
                    })];
            });
        });
    };
    /**
     * Triggers impressions
     *
     * The `triggerImpressions` method triggers multiple impressions for the given input. If provided
     * with Experience Short Uids, the impression is sent for the corresponding active variants in the manifest.
     * If provided wih Variant Aliases instead, it will use the Experience Short UID and Variant Short UID in the
     * alias to trigger impressions, without referring to the manifest.
     *
     * Call this method whenever an experience is shown to the user.
     *
     * Call this method after initializing the SDK.
     *
     * @param triggerImpressionOptions - Trigger Impression Options, one of
     * `experienceShortUids` or `aliases` must be given.
     * @example
     * ```ts
     * const experienceShortUids = ['a', 'b'];
     * await sdk.triggerImpressions({ experienceShortUids });
     * await sdk.triggerImpressions({ aliases: ['cs_personalize_a_0', 'cs_personalize_b_1'] });
     * ```
     */
    Sdk.prototype.triggerImpressions = function (triggerImpressionOptions) {
        return __awaiter(this, void 0, void 0, function () {
            var experienceAndVariants, impressions;
            var _a;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!triggerImpressionOptions.aliases && !triggerImpressionOptions.experienceShortUids) {
                            throw new Error('Cannot trigger impression without options');
                        }
                        if (triggerImpressionOptions.aliases) {
                            experienceAndVariants = triggerImpressionOptions.aliases.map(this.getExperienceAndVariantFromAlias);
                        }
                        else {
                            experienceAndVariants = triggerImpressionOptions.experienceShortUids.map(function (experienceShortUid) { return ({
                                experienceShortUid: experienceShortUid,
                                variantShortUid: _this.getActiveVariant(experienceShortUid)
                            }); }, this); // pass 'this' context to the map function to access the getActiveVariant method in minified file
                        }
                        impressions = experienceAndVariants
                            .filter(function (experienceAndVariant) { return !!experienceAndVariant.variantShortUid; })
                            .map(function (_a) {
                            var experienceShortUid = _a.experienceShortUid, variantShortUid = _a.variantShortUid;
                            return {
                                experienceShortUid: experienceShortUid,
                                variantShortUid: variantShortUid,
                                type: event_ingestion_type_1.IngestEventType.IMPRESSION
                            };
                        });
                        return [4 /*yield*/, (0, cross_fetch_1["default"])("".concat(this._edgeApiURL, "/events"), {
                                method: 'POST',
                                headers: (_a = {},
                                    _a[PROJECT_UID_HEADER_KEY] = this._projectUid,
                                    _a[USER_UID_HEADER_KEY] = this._userId,
                                    _a[X_USER_AGENT] = X_USER_AGENT_VALUE,
                                    _a),
                                body: JSON.stringify(impressions)
                            })];
                    case 1:
                        _b.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Trigger an event
     *
     * An event is any significant action performed by the user such as clicking on the CTA,
     * scrolling to the end of the page, etc. The event is triggered by passing an `eventKey`,
     * which is used to create the event in the Personalize Management console. Triggering events
     * allow you to measure conversions for an A/B test.
     *
     * Call this method after initializing the SDK.
     *
     * @param eventKey - The key to trigger an event for. Ensure that an event with this key is
     * created in Personalize.
     * @example
     * ```ts
     * await sdk.triggerEvent('clickCTA');
     * ```
     */
    Sdk.prototype.triggerEvent = function (eventKey) {
        return __awaiter(this, void 0, void 0, function () {
            var customEvent;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        customEvent = {
                            eventKey: eventKey,
                            type: event_ingestion_type_1.IngestEventType.CUSTOM_EVENT
                        };
                        return [4 /*yield*/, (0, cross_fetch_1["default"])("".concat(this._edgeApiURL, "/events"), {
                                method: 'POST',
                                headers: (_a = {},
                                    _a[PROJECT_UID_HEADER_KEY] = this._projectUid,
                                    _a[USER_UID_HEADER_KEY] = this._userId,
                                    _a[X_USER_AGENT] = X_USER_AGENT_VALUE,
                                    _a),
                                body: JSON.stringify([customEvent])
                            })];
                    case 1:
                        _b.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Set custom user attributes
     *
     * User attributes are key-value pairs describing user traits. To use the set attributes,
     * you must create the attribute with the same key in the Personalize Management console.
     *
     * This method can only be called after initializing the SDK.
     *
     * @param clientAttributes - Attributes set on the client. Must be an object.
     * @example
     * ```ts
     * await sdk.set({ age: 20 });
     * ```
     */
    Sdk.prototype.set = function (clientAttributes) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.setUserAttributesOnEdge(clientAttributes)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Sets the `userId`
     *
     * The `userId` is used to identify the current user. It is automatically generated at the edge
     * when initializing the SDK for the first time.
     *
     * Use this method when you want to identify the current user with an externally managed user ID.
     * For e.g., you can use this when an anonymous user logs in.
     *
     * When an anonymous user logs in, you may also want to preserve all the user attributes tracked thus far.
     * The option `preserveUserAttributes` is available for the same, which merges all the attributes from
     * the current user.
     *
     * The `setUserId` method also drops a cookie to track the user ID in browser environments. The cookie
     * is called `cs-personalize-user-id`.
     *
     * Call this method before or after initializing the SDK.
     *
     * @param userId - The new user ID for the current user
     * @param options - Pass options to preserve user attributes
     * @example
     * ```ts
     * await sdk.setUserId(newUserId, { preserveUserAttributes: true });
     * ```
     */
    Sdk.prototype.setUserId = function (userId, options) {
        return __awaiter(this, void 0, void 0, function () {
            var source_user_uid, mergeUserAttrsBody;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        source_user_uid = this._userId;
                        this._userId = userId;
                        if (this._isBrowserEnvironment) {
                            document.cookie = this.generateCookie(USER_UID_COOKIE_KEY, userId, {
                                days: USER_UID_COOKIE_MAX_AGE,
                                path: '/'
                            });
                        }
                        if (!(source_user_uid && (options === null || options === void 0 ? void 0 : options.preserveUserAttributes))) return [3 /*break*/, 2];
                        mergeUserAttrsBody = {
                            sourceUserUid: source_user_uid,
                            targetUserUid: userId
                        };
                        return [4 /*yield*/, (0, cross_fetch_1["default"])("".concat(this._edgeApiURL, "/user-attributes/actions/merge"), {
                                method: 'POST',
                                headers: (_a = {},
                                    _a[PROJECT_UID_HEADER_KEY] = this._projectUid,
                                    _a[X_USER_AGENT] = X_USER_AGENT_VALUE,
                                    _a),
                                body: JSON.stringify(mergeUserAttrsBody)
                            })];
                    case 1:
                        _b.sent();
                        _b.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Get the current `userId`
     *
     * Call this method after initializing the SDK.
     *
     * @returns the `userId`, else returns undefined if no `userId` is present
     */
    Sdk.prototype.getUserId = function () {
        return this._userId;
    };
    /**
     * Check for the active variant in the given experience
     *
     * Given an experience short UID, return the activated variant short UID for that experience.
     *
     * This method can only be called after initializing the SDK.
     *
     * @param experienceShortUid - Experience Short UID. Can be obtained from the
     * Personalize Management Console in the experience details page, or from a variant alias.
     * @returns The short UID of the active variant, if no variant is active, returns `null`.
     * @example
     * ```ts
     * const activeVariant = sdk.getActiveVariant(experienceShortUid);
     * ```
     */
    Sdk.prototype.getActiveVariant = function (experienceShortUid) {
        var _a;
        return (_a = this._manifestData.experiences.find(function (manifestExperience) { return manifestExperience.shortUid === experienceShortUid; })) === null || _a === void 0 ? void 0 : _a.activeVariantShortUid;
    };
    /**
     * Adds `userId` and manifest as set-cookie headers to the response
     *
     * `addStateToResponse` is a useful helper method that takes a response object
     * and adds cookies to track the user. It also adds a cookie to track the current manifest
     * for the user.
     *
     * The helper is typically used in an edge function to track the user state. When used this way,
     * it also allows the Personalize SDK to be initialized in the browser without making a
     * call to fetch the manifest.
     *
     * Call this method after initializing the SDK.
     *
     * @param response - A Web standard response object
     * @example
     * ```ts
     * await sdk.addStateToResponse(response);
     * ```
     */
    Sdk.prototype.addStateToResponse = function (response) {
        var userIdCookie = this.generateCookie(USER_UID_COOKIE_KEY, this._userId, {
            days: USER_UID_COOKIE_MAX_AGE,
            path: '/'
        });
        var manifestCookie = this.generateCookie(MANIFEST_COOKIE_KEY, JSON.stringify(this._manifestData), { path: '/' });
        response.headers.append(SET_COOKIE_HEADER_KEY, userIdCookie);
        response.headers.append(SET_COOKIE_HEADER_KEY, manifestCookie);
    };
    /**
     * Gets the active variants
     *
     * Returns a key-value pair of the active experience and variant short UIDs.
     *
     * An inactive experience will have `null` in the corresponding value.
     *
     * Call this method after initializing the SDK.
     *
     * @returns The active variants
     * @example
     * ```ts
     * sdk.getVariants(); // {a: 0, b: 1}
     * ```
     */
    Sdk.prototype.getVariants = function () {
        return this._manifestData.experiences.reduce(function (activeVariants, manifestExperience) {
            var _a;
            return __assign(__assign({}, activeVariants), (_a = {}, _a[manifestExperience.shortUid] = manifestExperience.activeVariantShortUid, _a));
        }, {});
    };
    /**
     * Gets the active variant aliases
     *
     * Returns a list of active experiences as variant aliases. Variant aliases are used by
     * Personalize to identify a CMS variant. The aliases can be passed to the CMS Delivery API
     * to fetch the personalized variant of content entries.
     *
     * The order in the list reflects the priority of the variants. The aliases higher in the list
     * have higher priority.
     *
     * Call this method after initializing the SDK.
     *
     * @returns The active variant aliases
     * @example
     * ```ts
     * sdk.getVariantAliases(); // ['cs_personalize_a_0', 'cs_personalize_b_1']
     * ```
     */
    Sdk.prototype.getVariantAliases = function () {
        return this.getCompactAliases().map(function (alias) { return "".concat(CS_PERSONALIZE_PREFIX, "_").concat(alias); });
    };
    /**
     * Gets the active variant param
     *
     * Returns an opaque variant parameter that can be easily passed in the URL.
     *
     * The parameter is essentially a comma separated list of active experience / variant short UIDs.
     *
     * The variant param can be passed as a query parameter in the URL to send the list of variants
     * activated for the current user. This helper method is typically used in edge functions.
     *
     * Call this method after initializing the SDK.
     *
     * @returns The active variant param
     * @example
     * ```ts
     * sdk.getVariantParam(); // 'a_0,b_1'
     * ```
     */
    Sdk.prototype.getVariantParam = function () {
        return this.getCompactAliases().join(',');
    };
    /**
     * Convert a variant param to a list of variant aliases
     *
     * Returns alist of variant aliases from a variant param.
     *
     * The variant param can be passed as a query parameter in the URL to send the list of variants
     * activated for the current user. This helper method is typically used in server side code to decode
     * the variant parameter as a list of variant aliases that can be passed to the CMS Delivery API.
     *
     * @param variantParam - The variant param generated by `sdk.getVariantParam()`
     * @returns The list of variant aliases
     * @example
     * ```ts
     * const variantParam = sdk.getVariantParam(); // 'a_0,b_1'
     * sdk.variantParamToVariantAliases(variantParam); // ['cs_personalize_a_0', 'cs_personalize_b_1']
     * ```
     */
    Sdk.prototype.variantParamToVariantAliases = function (variantParam) {
        return variantParam.split(',').map(function (alias) { return "".concat(CS_PERSONALIZE_PREFIX, "_").concat(alias); });
    };
    Sdk.prototype.getCompactAliases = function () {
        return this._manifestData.experiences
            .filter(function (experience) { return experience.activeVariantShortUid !== null; })
            .map(function (experience) { return "".concat(experience.shortUid, "_").concat(experience.activeVariantShortUid); });
    };
    Sdk.prototype.getExperienceAndVariantFromAlias = function (alias) {
        var _a = alias.split('_'), experienceShortUid = _a[2], _variantShortUid = _a[3];
        var variantShortUid = _variantShortUid === 'null' ? null : _variantShortUid;
        return { experienceShortUid: experienceShortUid, variantShortUid: variantShortUid };
    };
    Sdk.prototype.generateCookie = function (name, value, options) {
        var secondsInADay = 86400;
        var cookieString = "".concat(name, "=").concat(value);
        if (options === null || options === void 0 ? void 0 : options.days) {
            cookieString = "".concat(cookieString, "; Max-Age=").concat(options.days * secondsInADay);
        }
        if (options === null || options === void 0 ? void 0 : options.path) {
            cookieString = "".concat(cookieString, "; Path=").concat(options.path);
        }
        return cookieString;
    };
    Sdk.prototype.setUserAttributesOnEdge = function (userAttributes) {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, (0, cross_fetch_1["default"])("".concat(this._edgeApiURL, "/user-attributes"), {
                            method: 'PATCH',
                            headers: (_a = {},
                                _a[PROJECT_UID_HEADER_KEY] = this._projectUid,
                                _a[USER_UID_HEADER_KEY] = this._userId,
                                _a[X_USER_AGENT] = X_USER_AGENT_VALUE,
                                _a),
                            body: JSON.stringify(userAttributes)
                        })];
                    case 1: return [2 /*return*/, _b.sent()];
                }
            });
        });
    };
    return Sdk;
}());
exports.Sdk = Sdk;
