// @contentstack/personalize-edge-sdk/dist/sdk.spec.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var cross_fetch_1 = __importDefault(require("cross-fetch"));
var sdk_1 = require("./sdk");
jest.mock('cross-fetch');
var manifestResponse = {
    experiences: [
        {
            shortUid: '0',
            activeVariantShortUid: '0'
        },
        {
            shortUid: '1',
            activeVariantShortUid: '1'
        },
        {
            shortUid: '2',
            activeVariantShortUid: null
        },
    ]
};
var mockCookie = '';
var xUserAgentHeader = {
    'X-User-Agent': 'personalize-edge-sdk-js/0.0.0-development'
};
function initializeDocument() {
    Object.defineProperty(document, 'cookie', {
        get: jest.fn(function () { return mockCookie; }),
        set: jest.fn(function (newCookie) {
            var cookieName = newCookie.split('=')[0];
            var cookieValue = newCookie.split('=')[1].split(';')[0];
            var existingCookie = mockCookie.split('; ').find(function (cookie) {
                return cookie.split('=')[0] === cookieName;
            });
            if (existingCookie) {
                mockCookie = mockCookie.replace(existingCookie, "".concat(cookieName, "=").concat(cookieValue));
            }
            else {
                mockCookie += (mockCookie ? '; ' : '') + newCookie;
            }
        })
    });
}
describe('SDK', function () {
    var projectUid = 'projectUid';
    var edgeApiURL = 'mock-edge-api-url';
    var mockUserId = 'mock-user-id';
    var sdk;
    var manifestFetchResult;
    beforeEach(function () {
        document = {};
        initializeDocument();
        jest.clearAllMocks();
        manifestFetchResult = {
            ok: true,
            status: 200,
            headers: { get: jest.fn().mockReturnValue('randomly-generated-user-uid') },
            json: jest.fn().mockResolvedValue(manifestResponse)
        };
        cross_fetch_1["default"].mockClear().mockResolvedValue(manifestFetchResult);
        sdk = new sdk_1.Sdk(projectUid, manifestResponse, edgeApiURL, mockUserId, false, 'variant');
    });
    afterEach(function () {
        mockCookie = '';
    });
    describe('set user id', function () {
        it('sets user id as a cookie if running in browser environment', function () {
            sdk = new sdk_1.Sdk(projectUid, manifestResponse, edgeApiURL, mockUserId, true, 'variant');
            var mockCustomUserUid = 'mock-custom-user-uid';
            sdk.setUserId(mockCustomUserUid);
            expect(document.cookie).toBe("cs-personalize-user-uid=".concat(mockCustomUserUid, "; Max-Age=31536000; Path=/"));
        });
        it('sets user id only in memory if not running in browser environment', function () { return __awaiter(void 0, void 0, void 0, function () {
            var customUserId;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        document = undefined;
                        customUserId = 'custom_user_id';
                        return [4 /*yield*/, sdk.setUserId(customUserId)];
                    case 1:
                        _a.sent();
                        expect(sdk.getUserId()).toEqual(customUserId);
                        return [2 /*return*/];
                }
            });
        }); });
        it('merges the user attributes when preserveUserAttributes is true', function () { return __awaiter(void 0, void 0, void 0, function () {
            var mockTargetUserUid;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        sdk = new sdk_1.Sdk(projectUid, manifestResponse, edgeApiURL, mockUserId, true, 'variant');
                        mockTargetUserUid = 'target-user-uid';
                        return [4 /*yield*/, sdk.setUserId(mockTargetUserUid, {
                                preserveUserAttributes: true
                            })];
                    case 1:
                        _a.sent();
                        expect(sdk.getUserId()).toEqual(mockTargetUserUid);
                        expect(cross_fetch_1["default"]).toHaveBeenCalledWith('mock-edge-api-url/user-attributes/actions/merge', {
                            method: 'POST',
                            headers: __assign({ 'X-Project-Uid': projectUid }, xUserAgentHeader),
                            body: JSON.stringify({
                                sourceUserUid: 'mock-user-id',
                                targetUserUid: mockTargetUserUid
                            })
                        });
                        expect(document.cookie).toBe("cs-personalize-user-uid=".concat(mockTargetUserUid, "; Max-Age=31536000; Path=/"));
                        return [2 /*return*/];
                }
            });
        }); });
    });
    describe('get user id', function () {
        it('returns the user id from memory', function () {
            var mockCustomUserUid = 'mock-custom-user-uid';
            sdk.setUserId(mockCustomUserUid);
            var userId = sdk.getUserId();
            expect(userId).toBe(mockCustomUserUid);
        });
    });
    describe('set user attributes', function () {
        it('sets the user attributes at edge', function () { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, sdk.set({
                            name: 'Eclipse'
                        })];
                    case 1:
                        _a.sent();
                        expect(cross_fetch_1["default"]).toBeCalledWith('mock-edge-api-url/user-attributes', {
                            method: 'PATCH',
                            headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'mock-user-id' }, xUserAgentHeader),
                            body: JSON.stringify({
                                name: 'Eclipse'
                            })
                        });
                        return [2 /*return*/];
                }
            });
        }); });
    });
    describe('get active variant from experience', function () {
        it('returns active variant short uid if variant is active', function () {
            var expectedShortUid = '0';
            var result = sdk.getActiveVariant('0');
            expect(result).toBe(expectedShortUid);
        });
        it('returns null if active variant is null for given experience in manifest', function () {
            var result = sdk.getActiveVariant('2');
            expect(result).toBeNull();
        });
        it('returns undefined if given experience short uid not found in manifest', function () {
            var result = sdk.getActiveVariant('3');
            expect(result).toBeUndefined();
        });
    });
    describe('trigger impressions', function () {
        it('triggers the impression', function () { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, sdk.triggerImpression('1')];
                    case 1:
                        _a.sent();
                        expect(cross_fetch_1["default"]).toHaveBeenNthCalledWith(1, 'mock-edge-api-url/events', {
                            method: 'POST',
                            headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'mock-user-id' }, xUserAgentHeader),
                            body: JSON.stringify([
                                {
                                    experienceShortUid: '1',
                                    variantShortUid: '1',
                                    type: 'IMPRESSION'
                                },
                            ])
                        });
                        return [2 /*return*/];
                }
            });
        }); });
        it('triggers multiple impressions', function () { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, sdk.triggerImpressions({ experienceShortUids: ['0', '1'] })];
                    case 1:
                        _a.sent();
                        expect(cross_fetch_1["default"]).toHaveBeenCalledWith('mock-edge-api-url/events', {
                            method: 'POST',
                            headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'mock-user-id' }, xUserAgentHeader),
                            body: JSON.stringify([
                                {
                                    experienceShortUid: '0',
                                    variantShortUid: '0',
                                    type: 'IMPRESSION'
                                },
                                {
                                    experienceShortUid: '1',
                                    variantShortUid: '1',
                                    type: 'IMPRESSION'
                                },
                            ])
                        });
                        return [2 /*return*/];
                }
            });
        }); });
        it('triggers multiple impressions only for active variants', function () { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, sdk.triggerImpressions({
                            experienceShortUids: ['0', '1', '2']
                        })];
                    case 1:
                        _a.sent();
                        expect(cross_fetch_1["default"]).toHaveBeenCalledWith('mock-edge-api-url/events', {
                            method: 'POST',
                            headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'mock-user-id' }, xUserAgentHeader),
                            body: JSON.stringify([
                                {
                                    experienceShortUid: '0',
                                    variantShortUid: '0',
                                    type: 'IMPRESSION'
                                },
                                {
                                    experienceShortUid: '1',
                                    variantShortUid: '1',
                                    type: 'IMPRESSION'
                                },
                            ])
                        });
                        return [2 /*return*/];
                }
            });
        }); });
        it('triggers impressions with aliases', function () { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, sdk.triggerImpressions({ aliases: sdk.getVariantAliases() })];
                    case 1:
                        _a.sent();
                        expect(cross_fetch_1["default"]).toHaveBeenCalledWith('mock-edge-api-url/events', {
                            method: 'POST',
                            headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'mock-user-id' }, xUserAgentHeader),
                            body: JSON.stringify([
                                {
                                    experienceShortUid: '0',
                                    variantShortUid: '0',
                                    type: 'IMPRESSION'
                                },
                                {
                                    experienceShortUid: '1',
                                    variantShortUid: '1',
                                    type: 'IMPRESSION'
                                },
                            ])
                        });
                        return [2 /*return*/];
                }
            });
        }); });
        it('throws an error if neither aliases nor shortUids are provided', function () { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                expect(function () { return __awaiter(void 0, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0: return [4 /*yield*/, sdk.triggerImpressions({})];
                            case 1:
                                _a.sent();
                                return [2 /*return*/];
                        }
                    });
                }); }).rejects.toThrow('Cannot trigger impression without options');
                return [2 /*return*/];
            });
        }); });
    });
    describe('trigger event', function () {
        it('make call to events api when trigger event method is called', function () { return __awaiter(void 0, void 0, void 0, function () {
            var eventKey;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        eventKey = 'event_1';
                        return [4 /*yield*/, sdk.triggerEvent(eventKey)];
                    case 1:
                        _a.sent();
                        expect(cross_fetch_1["default"]).toBeCalledTimes(1);
                        expect(cross_fetch_1["default"]).toHaveBeenCalledWith('mock-edge-api-url/events', {
                            method: 'POST',
                            headers: __assign({ 'X-Project-Uid': projectUid, 'X-CS-Personalize-User-Uid': 'mock-user-id' }, xUserAgentHeader),
                            body: JSON.stringify([
                                {
                                    eventKey: eventKey,
                                    type: 'EVENT'
                                },
                            ])
                        });
                        return [2 /*return*/];
                }
            });
        }); });
    });
    describe('add state to response', function () {
        it('sets the userId and manifest in the set-cookie header', function () {
            var mockManifest = JSON.stringify(manifestResponse);
            var mockResponse = {
                headers: {
                    set: jest.fn(),
                    append: jest.fn()
                }
            };
            sdk.addStateToResponse(mockResponse);
            expect(mockResponse.headers.append).toHaveBeenNthCalledWith(1, 'Set-Cookie', 'cs-personalize-user-uid=mock-user-id; Max-Age=31536000; Path=/');
            expect(mockResponse.headers.append).toHaveBeenNthCalledWith(2, 'Set-Cookie', "cs-personalize-manifest=".concat(mockManifest, "; Path=/"));
        });
    });
    describe('get variants', function () {
        it('returns the active variants from manifest', function () {
            var result = sdk.getVariants();
            expect(result).toStrictEqual({
                '0': '0',
                '1': '1',
                '2': null
            });
        });
    });
    describe('get experiences', function () {
        it('returns the active experiences from the manifest', function () {
            var result = sdk.getExperiences();
            expect(result).toStrictEqual(manifestResponse.experiences);
        });
    });
    describe('alias management', function () {
        it('encodes variant aliases from the manifest', function () {
            var aliases = sdk.getVariantAliases();
            expect(aliases).toEqual(['cs_personalize_0_0', 'cs_personalize_1_1']);
        });
        it('encodes aliases as a compact query parameter', function () {
            var param = sdk.getVariantParam();
            expect(param).toEqual('0_0,1_1');
        });
        it('converts variant param to aliases', function () {
            var variantParam = '0_0,1_1';
            expect(sdk.variantParamToVariantAliases(variantParam)).toEqual([
                'cs_personalize_0_0',
                'cs_personalize_1_1',
            ]);
        });
    });
});
